FF: Make pyo parsing of devices more robust

guewen opened the pull request:
The methods to get input/output sound devices info in the pyo backend
rely on a parsing of the output of `pyo.pa_list_devices()`. It splits
the fields (name, host api index, latency, default sr) by splitting
on commas. This is fragile as the name can contain commas (e.g.
"Logitech USB Headset: Audio (hw:1,0)"), which leads to `IndexError:
list index out of range` error.

`pyo` provides a method [`pa_get_devices_infos()`](https://github.com/belangeo/pyo/blob/274ed61d6664d44d23416957e781b8afbb559589/src/engine/pyomodule.c#L139) that returns 2
dictionaries for input and output devices, almost exactly in the
expected format as returned by the current implementation of
`_query_devices()`. The only differences are in the fields values: no
units ("Hz", "s"), do not split the name to exclude the content after a
colon and the delay has many digits (vs 6 in the initial
implementation).

This commit switches to using `pa_get_devices_infos()` with a small
adaptation of the format to keep the same output (but the name) as the
previous implementation was doing.

:warning:  Behavior change:

The name of a sound device is now more precise, e.g.

Before: "Logitech USB Headset"
After: "Logitech USB Headset: Audio (hw:1,0)"

It makes the identifier more unique, as it contains the hardware position,
for example if we have several identical headsets connected.

Fixes #3365

guewen commented:
I'm not sure if I should pick BF or FF, please let me know, I'll gladly update my commit.
Also, as a new contributor, I may miss context or use cases, in such case, please do not mind to tell me :)

peircej reviewed (approved):
Thanks @guewen 
The BF/FF distinction is interesting here. This is a fix to existing released code (so normally BF) but my guess is that is could involve a behaviour change - the name of the audio card being returned might end up being different in the new more robust method, and that might cause existing experiment to stop working (exp can't find the audio card it was told to use?). So technically this is a BF but I might squahs it to FF so that it doesn't get included until the next major release (2021.1)

peircej commented:
> Thanks @guewen
> The BF/FF distinction is interesting here. This is a fix to existing released code (so normally BF) but my guess is that is could involve a behaviour change - the name of the audio card being returned might end up being different in the new more robust method, and that might cause existing experiment to stop working (exp can't find the audio card it was told to use?). So technically this is a BF but I might squahs it to FF so that it doesn't get included until the next major release (2021.1)

The question is whether we think the returned names will exactly match

peircej commented:
@guewen how urgent is this as a release for you? Do you have others that depend on it or are you OK because fixed on your local install?

guewen commented:
Thanks for the fast review!



> The question is whether we think the returned names will exactly match

AFAIU, the name should be identical, because both in `pa_get_devices_infos()` and `pa_list_devices()`, it comes from the same place (http://files.portaudio.com/docs/v19-doxydocs/structPaDeviceInfo.html).

That being said, I can't pretend to know enough and I have no guarantee that the name can't change (also on other systems such as windows), so I'm 100% fine with taking the safe way and include this in the next major release only. 

> Do you have others that depend on it or are you OK because fixed on your local install?

I'm fine with keeping this on my local install until the next major release.

guewen commented:
BTW it's worth checking what is the desired behavior as we put it as FF.

When we have a name such as `"Logitech USB Headset: Audio (hw:1,0)"`, the previous implementation was keeping only `Logitech USB Headset`, but re-reading the code, it's hard to know if it was the intent or only a side-effect of the manual parsing on `:` to split the keys and values (sounds like the second one).
I could change to keep the whole name intact. But it's probably not worth having to reconfigure exp in the wild if keeping only the first part was not an issue.

peircej commented:
> BTW it's worth checking what is the desired behavior as we put it as FF.

Yes, indeed. Let's allow potential change and save it for 2021.1. So, regarding your Q the advantage of keeping all of the name `"Logitech USB Headset: Audio (hw:1,0)"` might be that I guess that's identifying it more completely/uniquely. In your example above it looks like it's identifying its hardware position so this would be unique if you had two Logitech headsets connected, whereas if we split on the first `:` we could get a non-unique name. So my inclination would be to keep the whole name intact

guewen commented:
I'm not sure if I should pick BF or FF, please let me know, I'll gladly update my commit.



> I guess that's identifying it more completely/uniquely. In your example above it looks like it's identifying its hardware position so this would be unique if you had two Logitech headsets connected, whereas if we split on the first `:` we could get a non-unique name. So my inclination would be to keep the whole name intact

Very good point! I amended my commit to keep the name intact (and changed to FF).

peircej commented:
Thanks @guewen 
You're the perfect contributor! 😊
